feature/private cloud platform config (#798)
meshcloud/meshcloud-docs: docs/meshstack.cloudfoundry.index.md

title: Integration
---

meshStack supports integration with Cloud Foundry. Cloud Foundry is a PaaS platform which provides convenient application hosting capabilities to software and DevOps engineers.

meshStack supports project creation, configuration, access control, quota management and billing for Cloud Foundry.

## Integration Overview

> To enable integration with Cloud Foundry, Platform Operators configure one or multiple `meshPlatform`s of `PlatformType` Cloud Foundry in the [Platform Administration](./administration.platforms.md) in meshPanel.

Reviewer: ```suggestion
To enable integration with Cloud Foundry, Platform Operators configure one or multiple `Platform`s of `PlatformType` Cloud Foundry in the [Platform Administration](./administration.platforms.md) in meshPanel.
```

## Prerequisites 

Your Cloud Foundry must support the v3 Cloud Foundry API.
We have officially validated meshStack with support for Cloud Foundry `v3.108.0`.

### UAA Federated Users

meshStack will identify and assign users to roles in Cloud Foundry based on their `euid` (external user id) as described in [Identity Federation](meshstack.identity-federation.md#externally-provisioned-identities).

> You should set up your Cloud Foundry and UAA UAA so that they store the `euid` value in the `User.userName` field of the [UAA User object](https://docs.cloudfoundry.org/api/uaa/version/77.8.0/index.html#get) and `User.username` field of the [Cloud Foundry user object](http://v3-apidocs.cloudfoundry.org/version/3.164.0/index.html#users).

Reviewer: Just checking whether you wanted to mention UAA UAA twice.

## Integration Configuration

> > The recommended way to set up Cloud Foundry as a meshPlatform is via the public terraform [Cloud Foundry meshPlatform Module](https://github.com/meshcloud/terraform-cloudfoundry-meshplatform).

Reviewer: ```suggestion
> The recommended way to integrate Cloud Foundry as a platform in meshStack is via the public terraform [Cloud Foundry Platform Module](https://github.com/meshcloud/terraform-cloudfoundry-meshplatform).
```

Hmm now we did call the repo meshPlatform in the end. I leave it up to you whether you want to keep the meshPlatform naming but we did kind of decide to get rid of the meshNaming in our product.

Author: @Jelledb all of the platform integration tf modules are called `terraform-$provider-meshplatform`. The prefix `terraform-$provider` is mandatory by terraform registry conventions. If you want to call them differently, we'd need to migrate/rename all of them. If you want to get rid of that naming everywhere, that would require a dedicated backlog item.

### meshStack Admin User Accounts

meshStack requires two technical admin user accounts for integration with Cloud Foundry. We recommend your assign these 
users the following [Cloud Foundry roles](https://docs.cloudfoundry.org/concepts/roles.html) and [UAA scopes](https://docs.cloudfoundry.org/concepts/architecture/uaa.html#scopes).

- the `replicator` admin account must have permission to create orgs, spaces, groups and assign roles. This user requires the
  `Admin` role and additionally needs the `cloud_controller.admin`, `uaa.admin` and  `scim.read` scopes in UAA.
- the `metering` admin account is used to collect [metering data](./meshstack.cloudfoundry.metering.md). This requires the `Global Auditor`.

> The tenant [replication](./meshcloud.tenant.md) ensures spaces and orgs are created within the CF platform and appropriate permission rights are set when users access the CF platform. If a user's meshProject permissions are modified, meshStack updates the permissions for this user accordingly within the CF platform.

Reviewer: ```suggestion
The tenant [replication](./meshcloud.tenant.md) ensures spaces and orgs are created within the CF platform and appropriate permission rights are set when users access the CF platform. If a user's project permissions are modified, meshStack updates the permissions for this user accordingly within the CF platform.
```
